File: scala-advent-of-code/src/main/scala/advent/twenty_fifteen/Day14ReindeerOlympics.scala
package advent.twenty_fifteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

import scala.annotation.tailrec

case class Reindeer(name: String, speed: Int, flyTime: Int, restTime: Int)

class Day14ReindeerOlympics(filename: String, target: Int) extends DailyProblem[Int, Int] {
  private val pattern = "([A-Za-z]+) can fly ([0-9]+) km/s for ([0-9]+) seconds, but then must rest for ([0-9]+) seconds.".r

  private def parse(line: String): Reindeer = {
    line match {
      case pattern(name, speed, flyTime, restTime) => Reindeer(name, speed.toInt, flyTime.toInt, restTime.toInt)
    }
  }

  private val reindeer = FileUtilities.readFile(filename, parse)

  private def calculatePositions(time: Int): Map[String, Int] = reindeer.map(r => {
    val wholePeriods: Int = time / (r.flyTime + r.restTime)
    val remainingTime: Int = time % (r.flyTime + r.restTime)
    val flyTime: Int = Math.min(remainingTime, r.flyTime)
    r.name -> (wholePeriods * r.speed * r.flyTime + flyTime * r.speed)
  }).toMap

  private def calculatePoints(): Map[String, Int] = {
    @tailrec
    def iterate(time: Int, acc: Map[String, Int]): Map[String, Int] = {
      if (time >= target) {
        acc
      } else {
        val currentPositions = calculatePositions(time)
        val inFrontPositions = currentPositions.map(e => e._2).max
        val ahead = currentPositions.filter(e => e._2 == inFrontPositions)
        iterate(time + 1, acc.map(entry => if (ahead.contains(entry._1)) entry._1 -> (acc(entry._1) + 1) else entry))
      }
    }

    iterate(1, reindeer.map(r => r.name -> 0).toMap)
  }

  override val part1Answer: Int = calculatePositions(target).map(e => e._2).max
  override val part2Answer: Int = calculatePoints().map(e => e._2).max
}

File: java-advent-of-code/src/main/java/advent/twenty_fifteen/Day14ReindeerOlympics.java
package advent.twenty_fifteen;

import advent.common.DailyProblem;
import advent.twenty_fifteen.support.Reindeer;
import advent.utilities.FileUtilities;

import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Day14ReindeerOlympics implements DailyProblem<Integer, Integer> {
    private final int part1Answer;
    private final int part2Answer;

    public Day14ReindeerOlympics(final String filename, final int time) {
        List<Reindeer> reindeer = FileUtilities.readLines(filename, Reindeer::PARSE);

        Map<Reindeer, Integer> distanceTravelled = positionsAfterTime(time, reindeer);

        this.part1Answer = distanceTravelled.values().stream().max(Comparator.naturalOrder()).get();

        Map<Reindeer, Integer> points = new HashMap<>();
        for (int tick = 1; tick != time; tick++) {
            Map<Reindeer, Integer> distanceTravelledThisTick = positionsAfterTime(tick, reindeer);
            int max = distanceTravelledThisTick.values().stream().max(Comparator.naturalOrder()).get();
            for (Map.Entry<Reindeer, Integer> entry : distanceTravelledThisTick.entrySet()) {
                if (entry.getValue().intValue() == max) {
                    points.put(entry.getKey(), points.getOrDefault(entry.getKey(), 0) + 1);
                }
            }
        }

        this.part2Answer = points.values().stream().max(Comparator.naturalOrder()).get();
    }

    private Map<Reindeer, Integer> positionsAfterTime(int time, List<Reindeer> reindeer) {
        Map<Reindeer, Integer> distanceTravelled = new HashMap<>();

        for (Reindeer r : reindeer) {
            int wholePeriods = time / (r.getFlyPeriod() + r.getRestPeriod());
            int remainingTime = time % (r.getFlyPeriod() + r.getRestPeriod());
            int flyTime = Math.min(remainingTime, r.getFlyPeriod());
            distanceTravelled.put(r, wholePeriods * r.getSpeed() * r.getFlyPeriod() + flyTime * r.getSpeed());
        }

        return distanceTravelled;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}
